File: lib/scrp_conf.rb
class ConfigFlag
	public
	# flag init
	def initialize()
		@module_name = ""
		@module_type = ""
		@use_fifo_32 = false
		@use_fifo_8 = false
		@option_port = false
		@port_stack = []
		# @port_stack_size = 0
		@make_32_alw = false
		@alw32_stack = []
		# @alw32_stack_size = 0
		@make_8_alw = false
		@alw8_stack = []
		# @alw8_stack_size = 0
		@sub_module = false
		@sub_module_name = []
		@fi=""
	end
	

	def elem_ins()
		array = []
		i = 0
		while true
			l = @fi.gets.chomp

			if l == "}"
				@line = @line + 1
				break
			elsif l != "{" && l != "}"
				array[i] = l.delete("\t"+" ")
				# puts array[i]
				i = i + 1
			end
			@line = @line + 1
		end
		return array
	end

	def check_format(file)
		str = file.split(".")
		if str[1] != "scrp"
			puts "\n\nError! This file format is not supported by cReComop\n\n"
			exit(0)
		end
	end
	
	# state unit
	def set(dsl_file)
		check_format(dsl_file)
		@fi = open(dsl_file)
		@line = 1
		i_sub = 0
		i_alw32 = 0
		i_alw8 = 0
		i_op = 0
		while l = @fi.gets.to_s.chomp
			state = l.delete("{").split(" ")
			case state[0].to_s
				when "module_name" then
					@module_name = state[1]
					# puts @module_name
				when "module_type" then
					@module_type = state[1]
					# puts @module_type
				when "use_fifo_32" then
					@use_fifo_32 = true
					# puts "use_fifo_32"
				when "use_fifo_8" then
					@use_fifo_8 = true
					# puts "use_fifo_8"
				when "option_port" then
					@option_port = true
					# puts "option_port"
					@port_stack = elem_ins()
					i_op = i_op + 1
				when "make_32_alw" then
					@make_32_alw = true
					# puts "make_32_alw"
					@alw32_stack = elem_ins()
					i_alw32 = i_alw32 + 1
				when "make_8_alw" then
					@make_8_alw = true
					# puts "make_8_alw"
					@alw8_stack = elem_ins()
					i_alw8 = i_alw8 + 1
				when "sub_module_name" then
					@sub_module = true
					@sub_module_name[i_sub] = state[1]
					# puts "#{@sub_module_name[i_sub]}"
					i_sub = i_sub + 1
				when "end"
					break

				else
					puts "\nSyntax Error line #{@line} in #{dsl_file}"
					break
			end
			@line = @line + 1
		end
	end

	def get_module_name()
		return @module_name
	end
	def get_module_type()
		return @module_type
	end
	def get_use_fifo_32()
		if @use_fifo_32
			return true
		else
			return false
		end
	end
	def get_use_fifo_8()
		if @use_fifo_8
			return true
		else
			return false
		end
	end
	def get_option_port()
		if @option_port
			return @port_stack
		else
			return false
		end
	end
	def get_make_32_alw()
		if @make_32_alw
			return @alw32_stack
		else
			return false
		end
	end
	def get_make_8_alw()
		if @make_8_alw
			return @alw8_stack
		else
			return false
		end
	end
	def get_sub_module()
		if @sub_module
			return @sub_module_name
		else
			return false
		end
	end

end

# debug main
# flag = ConfigFlag.new()
# dsl_file = ARGV[0]

# flag.set(dsl_file)

# puts flag.get_module_name
# puts flag.get_module_type
# puts flag.get_use_fifo_32
# puts flag.get_use_fifo_8
# puts flag.get_option_port
# puts flag.get_make_32_alw
# puts flag.get_make_8_alw
# puts flag.get_sub_module

File: lib/scrp_conf.py
#!/usr/bin/python
# -*- coding: utf-8 -*-

class ConfigFlag(object):
	"""docstring for ConfigFlag"""
	def __init__(self,dsl_file):
		self.module_name = ""
		self.use_fifo_32 = False
		self.use_fifo_8 = False
		self.option_port = False
		self.port_stack = []
		self.make_32_alw = False
		self.alw32_stack = []
		self.make_8_alw = False
		self.alw8_stack = []
		self.sub_module = False
		self.sub_module_name = []
		# self.assign_target_module = []
		self.assign_port_stack = []
		self.fi=dsl_file
		self.line = 1
		self.w_cycle_32 = 0
		self.r_cycle_32 = 0
		self.w_cycle_8 = 0
		self.r_cycle_8 = 0
		self.connect = []
		self.make_reglist=False
		self.make_wirelist=False
		self.reg_list = []
		self.wire_list = []
		self.rw_condition_32 = False
		self.rw_condition_8 = False

	def elem_ins(self):
		array = []
		i = 0
		l = ""
		while True:
		 	l = self.fi.readline().rstrip()
			if "}" in l:
				self.line = self.line + 1
				break
			elif l != "{" and l != "}":
				array.append(l.translate(None,"\t "))
				i = i + 1
			self.line = self.line + 1
		return array

	def elem_ins_one(self):
		l = self.fi.readline().rstrip().translate(None,"\t")
		self.line = self.line + 1
		return l

	def check_format(self,filename):
		str_ = filename.split(".")
		if not str_[1] == "scrp":
			print "Error! This file format is not supported by cReComop"
			quit()

	def set(self,dsl_file):
		self.fi = open(dsl_file)
		i_sub = 0
		while True:
			line = self.fi.readline().translate(None,"{\t")
			if line == "\n":
				continue
			state = line.rstrip().split(" ")
			if "//" in state[0]:
				continue
			elif "module_name" == state[0]:
				self.module_name = state[1]
			elif "use_fifo_32" == state[0]:
				self.use_fifo_32 = True
			elif "use_fifo_8" == state[0]:
				self.use_fifo_8 = True
			elif "option_port" == state[0]:
				self.option_port = True
				self.port_stack = self.elem_ins()
			elif "make_32_alw" == state[0]:
				self.make_32_alw = True
				self.alw32_stack = self.elem_ins()
			elif "make_8_alw" == state[0]:
				self.make_8_alw = True
				self.alw8_stack = self.elem_ins()
			elif "sub_module_name" == state[0]:
				self.sub_module = True
				if  len(state)>2 and state[1].isdigit() == False and state[2].isdigit() == False:
					string = state[1] + " " + state[2]
					self.sub_module_name.append(string)
				else:
					print "Syntax Error near line %s in %s"%(self.line,dsl_file)
					quit()
			elif "assign_port" == state[0]:
				if state[2]!="" and state[2].isdigit()==False:
					self.connect = state[2]
				else:
					print "Syntax Error Specify cycle"
					quit()
				while True:
					l = self.fi.readline().rstrip()
					if l == "}":
						self.line = self.line + 1
						break
					elif l != "{" and l != "}":
						self.assign_port_stack.append(l.translate(None," \t"))
					self.line = self.line + 1
			elif "w_cycle_32" == state[0]:
				if state[1]!="" and state[1].isdigit():
					self.w_cycle_32 = state[1]
				else:
					print "Syntax Error Specify cycle"
					quit()
			elif "r_cycle_32" == state[0]:
				if state[1]!="" and state[1].isdigit():
					self.r_cycle_32 = state[1]
				else:
					print "Syntax Error Specify cycle"
					quit()
			elif "w_cycle_8" == state[0]:
				if state[1]!="" and state[1].isdigit():
					self.w_cycle_8 = state[1]
				else:
					print "Syntax Error Specify cycle"
					quit()
			elif "r_cycle_8" == state[0]:
				if state[1]!="" and state[1].isdigit():
					self.r_cycle_8 = state[1]
				else:
					print "Syntax Error Specify cycle"
					quit()
			elif "reg_list" == state[0]:
				self.make_reglist = True
				self.reg_list = self.elem_ins()
			elif "wire_list" == state[0]:
				self.make_wirelist = True
				self.wire_list = self.elem_ins()
			elif "rw_condition_32" == state[0]:
				self.rw_condition_32 = self.elem_ins_one()
				self.fi.readline()
			elif "rw_condition_8" == state[0]:
				self.rw_condition_8 = self.elem_ins_one()
				self.fi.readline()
			elif "end" in state[0]:
				break
			else:
				print "Syntax Error near line %s in %s"%(self.line,dsl_file)
				quit()
			self.line = self.line + 1

if __name__ == '__main__':
	
	filename = "sample.scrp"
	flag = ConfigFlag(filename)
	flag.check_format(filename)
	flag.set(filename)